### PR TITLE
mrc-5105: Add has_modifications flag to report list endpoint

### PR DESCRIPTION
**Pull request by r-ash**
To indicate when a particular report has modifications from the default branch.

We want this because in the UI when users view the list of reports, we want to have a checkbox to show only those reports which contain modifications (if the user has selected a branch other than the default branch)

This will always return `has_modifications` even if the hash this is called with the default branch or a hash which is on the default branch.

This gets the default branch from the `gert::git_remote_info("origin")`. This is the `HEAD branch` from the remote origin, which if I understand correctly should always be the GitHub default branch.

This gets the modified files using `git diff <from>...<to>` using the remote origin as the `<from>` branch. This is the same as what `orderly.server` is doing at the moment. I expect we could probably make an enhancement in `gert` to be able to call `gert::git_diff()` with a specified base ref but can do that later.

### CODE REVIEW

File: R/api.R [hunk 1]
```diff
@@ -33,17 +33,24 @@ root <- function() {
 
 ##' @porcelain 
 ##'   GET /report/list => json(report_list)
-##'   query hash :: string
+##'   query ref :: string
```
Comment (r-ash) [inline R/api.R:36]: I changed this to ref too because I think this is better description, it can work with a hash or a branch name

File: R/git.R [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  res <- system3(git, args)
+  if (check && !res$success) {
+    stop(sprintf("Error code %d running command:\n%s",
+                 res$code, paste0("  > ", res$output, collapse = "\n")))
+  }
+  res
+}
+
+
+git_ref_to_sha <- function(ref, repo = NULL, check = FALSE) {
```
Comment (richfitz) [inline R/git.R]: Is this not `gert::git_commit_id`?

File: R/git.R [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+git_get_default_branch <- function(repo = NULL) {
+  # This is assuming remote origin exists. We'll get an error if it
+  # doesn't. But this should be safe for us as we'll always have cloned
+  # this from GitHub.
+  origin <- gert::git_remote_info("origin", repo = repo)
+  origin$head
+}
+
+
+git_get_modified <- function(ref, base = NULL,
```
Comment (richfitz) [inline R/git.R:24]: we should make a ticket to patch gert for this perhaps

File: R/util_assert.R [hunk 4]
```diff
@@ -0,0 +1,23 @@
+assert_scalar <- function(x, name = deparse(substitute(x)), arg = name,
+                          call = NULL) {
+  if (length(x) != 1) {
+    cli::cli_abort(c("'{name}' must be a scalar",
+                     i = "{name} has length {length(x)}"),
+                   call = call, arg = arg)
+  }
+}
```
Comment (richfitz) [inline R/util_assert.R]: consider invisible return of x on all of these (not needed on assert_scalar_character as that's implicit) as that's often useful

**Review by richfitz — CHANGES_REQUESTED** [review]

File: R/git.R (hunk 2 above)
Comment (r-ash) [reply R/git.R]: It absolutely is yes, thanks!

File: R/git.R (hunk 3 above)
Comment (r-ash) [reply R/git.R:24]: Agreed, added a ticket for this https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-5126/Update-gert-to-be-able-to-get-diff-with-...-syntax

File: R/util_assert.R [hunk 5]
```diff
@@ -0,0 +1,25 @@
+assert_scalar <- function(x, name = deparse(substitute(x)), arg = name,
```
Comment (richfitz) [inline R/util_assert.R]: After your change you no longer need these :) but they're bound to come in useful lter

File: tests/testthat/test-util-assert.R [hunk 6]
```diff
@@ -0,0 +1,12 @@
+test_that("assert_scalar", {
+  expect_error(assert_scalar(NULL), "must be a scalar")
+  expect_error(assert_scalar(numeric(0)), "must be a scalar")
+  expect_error(assert_scalar(1:2), "must be a scalar")
```
Comment (richfitz) [inline tests/testthat/test-util-assert.R]:
```suggestion
  expect_error(assert_scalar(1:2), "must be a scalar")
  expect_no_error(assert_scalar(1))
```

**Review by richfitz — APPROVED** [review]

**Review by M-Kusumgar — APPROVED** [review]
LGTM!

**Pull request merged**